usecase/pokemon.go

```diff
@@ -0,0 +1,168 @@
+	GetPokemonsConcurrently(typeNumber string, items int, itemsPerWorker int) ([]model.Pokemon, *model.Error)
+	GetPokemon(pokemonId int) (model.Pokemon, *model.Error)
+	GetPokemonsFromExternalAPI() (*[]model.SinglePokeExternal, *model.Error)
+}
+
+func New(s csvservice.NewCsvService, h httpservice.NewHttpService) *PokemonUsecase {
+	return &PokemonUsecase{s, h}
+}
+
+func (us *PokemonUsecase) GetPokemons() ([]model.Pokemon, *model.Error) {
+	return us.csvService.GetPokemons()
+}
+
+func (us *PokemonUsecase) GetPokemon(pokemonId int) (model.Pokemon, *model.Error) {
+	return us.csvService.GetPokemon(pokemonId)
+}
+
+func calculatePoolSize(items int, itemsPerWorker int, totalPokemons int) int {
+	var poolSize int
+	if items%itemsPerWorker != 0 {
+		poolSize = int(math.Ceil(float64(items) / float64(itemsPerWorker)))
+	} else {
+		poolSize = int(items / itemsPerWorker)
+	}
+
+	// If we overpass the number of workers above the half of number
+	// of items it's gonna get into an infinit looop
+	if poolSize > (totalPokemons / 2) {
+		poolSize = totalPokemons / 2
+	}
+	return poolSize
+}
+
+func calculateMaxPokemons(totalPokemons int) int {
+	var maxPokemons int
+
+	if totalPokemons%2 == 0 {
+		maxPokemons = totalPokemons / 2
+	} else {
+		maxPokemons = totalPokemons/2 + 1
+	}
+	return maxPokemons
+}
+
+func (us *PokemonUsecase) GetPokemonsConcurrently(typeNumber string, items int, itemsPerWorker int) ([]model.Pokemon, *model.Error) {
+	pokemons, err := us.csvService.GetPokemons()
+
+	if err != nil {
+		return nil, err
+	}
+
+	filterType := model.TypeNumberFilter{
+		Even: "even",
+		Odd:  "odd",
+	}
+	totalPokemons := len(pokemons)
+	poolSize := calculatePoolSize(items, itemsPerWorker, totalPokemons)
+	maxPokemons := calculateMaxPokemons(totalPokemons)
+
+	values := make(chan int)
+	jobs := make(chan int, poolSize)
+	shutdown := make(chan struct{})
+
+	startIndex := 0
+	var limit int
+	limit = int(math.Ceil(float64(totalPokemons) / float64(poolSize)))
+	lastLimit := (totalPokemons % limit)
+
+	var wg sync.WaitGroup
+	wg.Add(poolSize)
+
+	for i := 0; i < poolSize; i++ {
+		go func(jobs <-chan int) {
+			for {
+				var id int
+				var limitRecalculated int
+				start := <-jobs
+
+				// We do need to iterate with the same limit every time.
+				// on the last cycle we use the leftovers of the division (modulus)
+				if limit+start >= totalPokemons && lastLimit != 0 { // lastLimit can be 0, take care of that
+					limitRecalculated = start + lastLimit
+				} else {
+					limitRecalculated = start + limit
+				}
+
+				for j := start; j < limitRecalculated; j++ {
+					id = pokemons[j].ID
+
+					select {
+					case values <- id:
+					case <-shutdown:
+						wg.Done()
+						return
+					}
+				}
+			}
+		}(jobs)
+	}
+
+	for i := 0; i < poolSize; i++ {
+		jobs <- startIndex
+		startIndex += limit
+	}
+	close(jobs)
+
+	var filteredPokemons []model.Pokemon = nil
+	bucket := make(map[int]int, totalPokemons+1)
+	for elem := range values {
```

Comment:
Shouldn't this validation be done within the worker itself? It could be that we get to this point with an empty or incomplete array of values, doesn't it?